### PR TITLE
GCS_MAVLink: tidy send_message function

### PR DESCRIPTION
**Pull request by peterbarker**

### CODE REVIEW

**Comment by peterbarker** [comment]
@OXINARF Does this seem clearer to you?

There shouldn't be any functionality change, but it is marginally more efficient in that we don't seq-scan the deferred message list if there's nothing currently queued, and we avoid a second seq-scan to insert the message into the deferred queue if we choose to defer.

Marked WIP because I need to stare at this a while longer; not a good routine to stuff up :-)

**Comment by magicrub** [comment]
What is this trying to solve?

**Comment by peterbarker** [comment]
@magicrub: readability issues, primarily.

It also removes a pitfall by using ARRAY_SIZE() in place of the magic final element as a flag value.  That array is actually also currently mis-sized as we will never queue MSG_RETRY_DEFERRED...

As mentioned earlier, it is also marginally more efficient.

It's marked WIP because it isn't actually FIXING any problem, and I wanted others input on whether it was a change worth making.

File: libraries/GCS_MAVLink/GCS.h [hunk 1]
```diff
@@ -74,7 +74,7 @@ enum ap_message {
     MSG_BATTERY_STATUS,
     MSG_AOA_SSA,
     MSG_LANDING,
-    MSG_RETRY_DEFERRED // this must be last
+    MSG_RETRY_DEFERRED
```
Comment (peterbarker) [inline libraries/GCS_MAVLink/GCS.h]: Ah, but with these patches that array doesn't *need* to be that size any more :-)

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 2]
```diff
@@ -733,48 +733,53 @@ void GCS_MAVLINK::send_message(enum ap_message id)
...
+        if (try_send_message(id)) {
+            // yay, we sent it!
+            return;
+        }
+    }
+
+    // we failed to send the message this time around, so try to defer:
+    if (num_deferred_messages == ARRAY_SIZE(deferred_messages)) {
+        // the defer buffer is full, discard this attempt to send.
+        // Note that the message *may* already be in the defer buffer
```
Comment (peterbarker) [inline libraries/GCS_MAVLink/GCS_Common.cpp]: Again, we currently size it so you can defer one of each message (and at 4bytes*~20 elements, *shrug*).  But it doesn't need to be that way.

**Comment by peterbarker** [comment]
@OXINARF `gcs().retry_deferred()` implemented.

**Comment by OXINARF** [comment]
> Ah, but with these patches that array doesn't need to be that size any more :-)

> Again, we currently size it so you can defer one of each message (and at 4bytes*~20 elements, shrug). But it doesn't need to be that way.

Ok, if the intention isn't to necessarily allow each message to be deferred once, then I would say that we scrap the MSG_LAST and just put a number in the array size. If on the other hand we want to keep MSG_LAST, then my previous comments still apply.

> gcs().retry_deferred() implemented.

Love it, I'm just confused why we need a `push_deferred_messages` that simply calls `retry_deferred`?

**Comment by peterbarker** [comment]
@OXINARF I hope it is `retry_deferred` calling `push_deferred_messages`... Basically so the frontend isn't having too intimate a knowledge as to what the backend is up to.  There may be other things a backend might be doing apart from this deferred-message-queue thing.  And yes, I did quarrel with myself over doing that :-)

As to the array size intention; no comment, so I don't know what the intent is or was.  One thing that did occur to me is that this could have developed from a bitmask of messages that needed to be resent.

Would be an interesting exercise to see what the most message types each vehicle will ever queue is - but I kind of figured not changing functionality is a good idea.  If we were to make changes I'd be looking at casting to a uint8_t for `deferred_messages[]`, saving 135 bytes of RAM.  And maybe adding a warning about dramatically increasing the enumeration size...

I think with the `retry_deferred()` thing this is actually looking worthwhile now.  Should we press on with this one?

**Comment by OXINARF** [comment]
> I hope it is `retry_deferred` calling `push_deferred_messages`...

Yes 😃 

> Basically so the frontend isn't having too intimate a knowledge as to what the backend is up to. 

OK, but the frontend is calling `retry_deferred`, I think that is enough. I guess you are thinking that having `send_message ` calling a method named `retry_deferred` is strange? It looks normal to me.
If I'm not clear: my proposal is to remove `push_deffered_messages` (by inlining it into `retry_deferred`).
Last time I push for this though, not that important.

> As to the array size intention; no comment, so I don't know what the intent is or was. One thing that did occur to me is that this could have developed from a bitmask of messages that needed to be resent.

Let me try to put my thinking here:

- if we have MSG_LAST and use it just for the `deferred_messages` array size, then we are declaring the array has space for each of enum values. And in this case, I think the code comments should be as I said before: MSG_LAST with a comment saying it should be last and the *may* word removed from the comment
- if we remove MSG_LAST and use a numeric value for the `deferred_messages` array size (even if the size is still the same as now), then we are declaring that array has that size but can change at any time. In this case the comments should be like you've put them.

I'll always push this one, I think it is a matter of code comments being coherent with the code.

> Would be an interesting exercise to see what the most message types each vehicle will ever queue is - but I kind of figured not changing functionality is a good idea. If we were to make changes I'd be looking at casting to a uint8_t for deferred_messages[], saving 135 bytes of RAM. And maybe adding a warning about dramatically increasing the enumeration size...

We can do that, we just need to add before or after the array:

`static_assert(MSG_LAST - 1 <= std::numeric_limits<uint8_t>::max(), "ap_message enum has values bigger than can fit in deferred_messages array type");`

Untested, but something like that should work.

> I think with the retry_deferred() thing this is actually looking worthwhile now. Should we press on with this one?

Yes, I think this is worthwhile now.

**Comment by peterbarker** [comment]
@OXINARF: yeah, inlining `push_deferred_messages` was tempting.  I kept it out because the frontend pushes things other than the deferred messages as part of `retry_deferred`  - it also pushes the statustext.  So `retry_deferred` is the generic "do deferred stuff", `push_deffered_mesages` refers specifically to the `ap_message` queue.  If we did want to do more stuff in `retry_deferred` we wouldn't want to do it as part of `send_message`.  I just revisited this to inline as you suggested and decided against it :-P

I've added a comment against `MSG_LAST` saying it must be last, and a comment explaining the size we use for the array.

I've left the "may" in - removing it would couple that and the sizing of the array done elsewhere, leading to a reasonable possibility the comment would become misleading later!

Let's tackle the ap_message/uint8_t change as a separate issue.

**Comment by OXINARF** [comment]
@peterbarker Can you please rebase this so I can look at it again? Thanks.

**Comment by peterbarker** [comment]
@OXINARF Done.

**Comment by OXINARF** [comment]
@peterbarker Please check https://github.com/ArduPilot/ardupilot/compare/master...OXINARF:review/6622. Final code is exactly the same, but:

- I've split the commits a bit differently (basically change to MSG_LAST is separated, with the last commit in the PR squashed into it)
- changed the commit messages just a bit

Give me your 👍 or 👎

**Comment by peterbarker** [comment]
@OXINARF :+1:

**Comment by OXINARF** [comment]
Merged my branch, thanks!

**Pull request closed without merging**